### DISCUSSION TITLE
CoreWCF application takes > 10 seconds to exit when running on Linux

### DISCUSSION THREAD

**Discussion by Sune1337** [CoreWCF/CoreWCF#1122] [Q&A] [answered]
I'm building a .Net 7 CoreWCF service, and when i exit this application (using ctrl-c) it takes > 10 seconds for the process to quit when i run on Linux.

On Windows, the application exits immediately.

In production i run the application in a Linux container, but i see the "slow exit" behavior also when running on the host os.
I've tested this on both Debian and Rocky linux.

Have anyone else seen this; or even better have a resolution for it?
When i run this in my docker-environment (using docker compose plugin) the docker logs an error "Container failed to exit within 10s of signal 15 - using the force" when the compose-service is shutdown.

I see this same behavior also when running the [sample service](https://github.com/CoreWCF/samples/tree/main/Basic/Binding/Basic/Http/Service).

**Comment by Sune1337** [comment]
I found some clues to this issue.

First; basicHttpBinding.OpenTimeout has a default value of 1 minute. If i ctrl-c the application before 1 minute has passed since i started it, the app will wait until 1 minute has passed and then exit.

Second; serviceHostBase.CloseTimeout has a default value of 10 seconds. If i ctrl-c the application, it will wait 10 seconds and then exit. (assuming the application has been running more than 1 minute so that the OpenTimeout has passed once)

To make it more clear which timeouts i'm referring to, this is the code to set them to 1 second so that the application exits faster.

```cs
var basicHttpBinding = new BasicHttpBinding();
// Default OpenTimeout = 1 minute.
basicHttpBinding.OpenTimeout = new TimeSpan(0, 0, 1);

app.UseServiceModel(serviceBuilder =>
{
    // Default CloseTimeout = 10 seconds.
    serviceBuilder.ConfigureAllServiceHostBase(a => a.CloseTimeout = new TimeSpan(0, 0, 1));
});
```

Since my services will only run on a LAN i think this will be an OK workaround for me. (If the timeouts mean what i think they do)

But i still think this is an incorrect behavior since on Windows it always exits fast even if the default timeouts are the same. Also; perhaps there are other timeouts that activate when the service is actually in use.

**Reply by mconnew** [reply]
Can you debug the app and break into it after you've pressed ctrl-c to see if there's a call stack is blocked waiting? The Open timeout should be meaningless for HTTP.

**Comment by Sune1337** [comment]
There's a few threads running:

![image](https://github.com/CoreWCF/CoreWCF/assets/7024113/28abed14-6be5-4149-ad8f-8f69c0c8ead1)

<details>
<summary>All call stacks</summary>
 @2721

 @2778
	WaitHandle.WaitMultiple()
	WaitHandle.WaitMultiple()
	WaitHandle.WaitAny()
	IOThreadTimer.WaitableTimer.WaitAny() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/bca178c28485aaedff7f9ca4c55eadb65721fe13928b9641a64ce97b17136/IOThreadTimer.cs:line 609
	IOThreadTimer.TimerManager.OnWaitCallback()
	IOThreadScheduler.ScheduledOverlapped.CallbackCore()
	IOThreadScheduler.ScheduledOverlapped.Callback()
	Thread.StartCallback()
	[Native to Managed Transition]

 @2741

 @2750

 @2743

.NET File Watcher @2745
	FileSystemWatcher.RunningInstance.TryReadEvent() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/fc28c1356cdf9dfd58c06ee3ff6f4ee3eacbe698925edd48f013e44af71ca6/FileSystemWatcher.Linux.cs:line 769
	FileSystemWatcher.RunningInstance.ProcessEvents() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/fc28c1356cdf9dfd58c06ee3ff6f4ee3eacbe698925edd48f013e44af71ca6/FileSystemWatcher.Linux.cs:line 536
	FileSystemWatcher.RunningInstance.<>c.<Start>b__14_0()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET Sockets @2752
	[Native frame]
	SocketAsyncEngine.EventLoop() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/2b4ec1d63cce835dfcbf8cc95f5a19ec1c94664867cff0ec4dcdd746e39e9e/SocketAsyncEngine.Unix.cs:line 176
	SocketAsyncEngine.<>c.<.ctor>b__14_0()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Gate @2747
	WaitHandle.WaitOneNoCheck() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/ddd9ab8cc6ed9c3162912913c88015b8249465abbf24a90917199e1a6d76/WaitHandle.cs:line 139
	WaitHandle.WaitOne()
	PortableThreadPool.GateThread.GateThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @3027
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/f2c3cf915285f1322ca331a75b76f458504989139e448c2cca5c6859d64e978a/LowLevelLifoSemaphore.cs:line 204
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @2746
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal() at /home/iperf/.config/JetBrains/Rider2023.1/resharper-host/SourcesCache/f2c3cf915285f1322ca331a75b76f458504989139e448c2cca5c6859d64e978a/LowLevelLifoSemaphore.cs:line 204
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @2748
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @2756
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

.NET ThreadPool Worker @2753
	[Native frame]
	LowLevelLifoSemaphore.WaitForSignal()
	LowLevelLifoSemaphore.Wait()
	PortableThreadPool.WorkerThread.WorkerThreadStart()
	Thread.StartCallback()
	[Native to Managed Transition]

Worker thread @3032
</details>

The one that i dug into to eventually start to experiment with different timeouts is @2778.
That's because i suspected something in CoreWCF and that's the only thread with CoreWCF in the call stack.
  (the exit is only slow if i call serviceBuilder.AddService + AddServiceEndpoint.
![image](https://github.com/CoreWCF/CoreWCF/assets/7024113/4fe9a4e2-7173-446e-8e09-113b8ceb3616)

**Comment by parus95** [comment, marked as answer]
Workaround:

add to Program.cs before `app.Run();`:

``` csharp
app.Services.GetRequiredService<IHostApplicationLifetime>().ApplicationStopping.Register(() =>
{
    foreach (var assemblyName in new[] { "CoreWCF.Http", "CoreWCF.Primitives" })
    {
        System.Reflection.Assembly.Load(assemblyName).GetType("CoreWCF.Runtime.IOThreadTimer")
            .GetMethod("KillTimers", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static)
            .Invoke(null, new object[] { });
    }
    app.Services.GetRequiredService<ILogger<object>>().LogInformation("CoreWCF timers killed");
});
```

**Comment by mconnew** [comment]
I created issue #1361 from this as it should be fixed. Thank you for working out the root cause, this takes a lot of the work out of being able to ship a change to address this. I'm going to close this discussion as it's understood, and there's a temporary workaround posted. Any further discussion can happen on the issue.
